allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -229,6 +229,12 @@ def _read(self, file_path: str) -> Iterable[Instance]:
         """
         raise NotImplementedError
 
+    def get_vocabulary(self) -> Optional[Vocabulary]:
+        """Returns the vocabulary used in the created instances. By default, this
+        returns `None`, which causes the vocabulary to be automatically discovered
```

Comment:
This isn't actually the behavior of the current code in `train.py`.  But we can assume that we fix that to match this description.  What I would like to wrap my head around is what happens in the following cases:

1. A glove vocabulary
2. A BERT QA model (no tag vocab needed)
3. A glove tagging model
4. A BERT tagging model

What do you imagine a person specifying where, and how does vocabulary creation happen?  Putting this method here is less scary than I originally thought, but I'm still somewhat skeptical that we really gain what you're hoping.  Outline what code / parameters a person has to specify in the above cases would help clarify that.

Reply:
1:
```Python
def get_vocabulary(self):
    # if we write a utility function
    return Vocabulary.from_glove(d=100)
    
    # if we don't write a utility function
    vocabulary = Vocabulary.empty()
    with open(cached_path("http://.../vocab.txt")) as f:
        vocabulary.add_tokens_to_namespace([t.strip() for t in f], "tokens")
    return vocabulary

    # or we could write a utility function for loading from txt files
    return Vocabulary.from_txt("http://.../vocab.txt")
```
This is assuming that Glove ships with a vocab.txt where each line contains a vocab item, and the order matches the order of the vectors.

2:
```Python
def get_vocabulary(self):
    return Vocabulary.from_transformers(self.transformer_model)
```

3:
```Python
def get_vocabulary(self):
    vocabulary = Vocabulary.from_glove(d=100)
    vocabulary.add_tokens_to_namespace(self.possible_tags, "tags")
    return vocabulary
```

4:
```Python
def get_vocabulary(self):
    vocabulary = Vocabulary.from_transformers(self.transformer_model)
    vocabulary.add_tokens_to_namespace(self.possible_tags, "tags")
    return vocabulary
```

How does the current mechanism make sure that in the case of Glove, we get the right indices for the right tokens?

Reply:
For case 1, we typically don't load the whole glove vocabulary, because that's massive.  We just load the embeddings for the tokens that we've seen, saving a ton of space.  You can't do that without looking through your instances and computing a vocabulary.

For cases 3 and 4, if you're saying that the reader has to hard-code a list of tags / labels, this seems problematic.  It means that I can't have a generic "classification csv file" dataset reader, because the reader would have to know the label set.